src/Components/Aspire.Azure.Messaging.WebPubSub/AspireWebPubSubExtensions.cs

```diff
@@ -0,0 +1,136 @@
+    /// <param name="hubName">The hub used for Web PubSub service.</param>
+    /// <param name="configureSettings">An optional method that can be used for customizing the <see cref="AzureMessagingWebPubSubSettings"/>. It's invoked after the settings are read from the configuration.</param>
+    /// <param name="configureClientBuilder">An optional method that can be used for customizing the <see cref="IAzureClientBuilder{WebPubSubServiceClient, WebPubSubServiceClientOptions}"/>.</param>
+    /// <remarks>Reads the configuration from "Aspire.Azure.Messaging.WebPubSub:{name}" section.</remarks>
+    /// <exception cref="InvalidOperationException">Thrown when neither <see cref="AzureMessagingWebPubSubSettings.ConnectionString"/> nor <see cref="AzureMessagingWebPubSubSettings.Endpoint"/> is provided.</exception>
+    public static void AddKeyedAzureWebPubSub(
+        this IHostApplicationBuilder builder,
+        string name,
+        string hubName,
+        Action<AzureMessagingWebPubSubSettings>? configureSettings = null,
+        Action<IAzureClientBuilder<WebPubSubServiceClient, WebPubSubServiceClientOptions>>? configureClientBuilder = null)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(name);
+
+        string configurationSectionName = WebPubSubComponent.GetKeyedConfigurationSectionName(name, DefaultConfigSectionName);
+
+        new WebPubSubComponent(hubName).AddClient(builder, configurationSectionName, configureSettings, configureClientBuilder, connectionName: name, serviceKey: name);
+    }
+
+    private sealed class WebPubSubComponent(string hubName) : AzureComponent<AzureMessagingWebPubSubSettings, WebPubSubServiceClient, WebPubSubServiceClientOptions>
+    {
+        protected override IAzureClientBuilder<WebPubSubServiceClient, WebPubSubServiceClientOptions> AddClient<TBuilder>(TBuilder azureFactoryBuilder, AzureMessagingWebPubSubSettings settings, string connectionName, string configurationSectionName)
+        {
+            return azureFactoryBuilder.RegisterClientFactory<WebPubSubServiceClient, WebPubSubServiceClientOptions>((options, cred) =>
+            {
+                var connectionString = settings.ConnectionString;
+                if (string.IsNullOrEmpty(connectionString) && settings.Endpoint == null)
+                {
+                    throw new InvalidOperationException($"A WebPubSubServiceClient could not be configured. Ensure valid connection information was provided in 'ConnectionStrings:{connectionName}' or specify a 'ConnectionString' or 'Endpoint' in the '{configurationSectionName}' configuration section.");
+                }
+
+                return !string.IsNullOrEmpty(connectionString) ?
+                    new WebPubSubServiceClient(connectionString, hubName, options) :
+                    new WebPubSubServiceClient(settings.Endpoint!, hubName, cred, options);
```

Comment:
The Blobs component checks if `cred` is null. Do we need to here?

https://github.com/dotnet/aspire/blob/b93520de6bccbdd59af5f57ed5f0d62aee921159/src/Components/Aspire.Azure.Storage.Blobs/AspireBlobStorageExtensions.cs#L114-L116

Reply:
no, webpubsub does not support null `cred` when the first parameter is Endpoint uri